genie-agent/src/main/java/com/netflix/genie/agent/execution/services/LaunchJobServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services;
+
+import com.netflix.genie.agent.execution.ExecutionContext;
+import com.netflix.genie.agent.execution.exceptions.JobLaunchException;
+import com.netflix.genie.agent.utils.PathUtils;
+import com.netflix.genie.common.dto.v4.JobSpecification;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Configures and launches a job sub-process using metadata passed through ExecutionContext.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+@Component
+@Lazy
+class LaunchJobServiceImpl implements LaunchJobService {
+
+    private final ExecutionContext executionContext;
+
+    LaunchJobServiceImpl(final ExecutionContext executionContext) {
+        this.executionContext = executionContext;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Process launchProcess() throws JobLaunchException {
+        final JobSpecification jobSpec = executionContext.getJobSpecification();
+
+        if (jobSpec == null) {
```

Comment:
So this assumes ExecutionContext is created before job spec is resolved and thus could be null. I think the execution context should either return `Optional<JobSpecification>` or we maybe add some sort of execution state flag into the context so we know it's already resolved?

Reply:
I think my point was more about the contract (API) between the `ExecutionContext` and its consumers. In this case when I look at your API definition you don't have the return marked as `@Nullable` and you aren't returning `Optional` so there's no real signal outside the class that this value could be null without looking at hte source or understanding the entire flow of the system. I think one of those two above options should be present on APIs so that we can trust class behavior as we consume them around the system

Reply:
Oh, I see. I'll mark all fields `@Nullable`